[tests] Add messages.py

jnewbery opened the pull request:
Second part of #11518.

Moves the primitive Bitcoin datastructures and message classes into their own module, and tidies up the mininode.py module.

- First commit is almost entirely move-only
- Second commit is mostly move-only, but also does a little tidying.

jnewbery committed:
[tests] Move test_framework Bitcoin primitives into separate module

mininode.py wildcard imports all names from primitives.py. This is
to avoid having to change all test scripts that import from mininode.py.

jnewbery committed:
[tests] Tidy up mininode.py module

Mostly move only. Adds a few extra comments.

fanquake commented:
One travis failure in net.py:
```
2017-11-10 01:10:17.553000 TestFramework.node0 (DEBUG): RPC successfully started
2017-11-10 01:10:17.555000 TestFramework.node1 (DEBUG): RPC successfully started
2017-11-10 01:10:18.043000 TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/test_framework/test_framework.py", line 120, in main
    self.run_test()
  File "/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/net.py", line 27, in run_test
    self._test_getnettotals()
  File "/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/net.py", line 54, in _test_getnettotals
    assert_equal(before['bytesrecv_per_msg']['pong'] + 32, after['bytesrecv_per_msg']['pong'])
  File "/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/test/functional/test_framework/util.py", line 38, in assert_equal
    raise AssertionError("not(%s)" % " == ".join(str(arg) for arg in (thing1, thing2) + args))
AssertionError: not(64 == 32)
2017-11-10 01:10:18.044000 TestFramework (INFO): Stopping nodes
2017-11-10 01:10:18.044000 TestFramework.node0 (DEBUG): Stopping node
```

MeshCollider commented:
@fanquake looks unrelated, restarted

laanwj commented:
Thanks for trying to clean things up, but I think we need more categorization here; bitcoin P2P message types are clearly a different kind of primitives than consensus data structures such as CTransaction CTXin etc. Only the second group is called "primitives" in the C++ code.

ryanofsky commented:
Maybe just call it something like serialize.py or messages.py.

jnewbery commented:
> I think we need more categorization here; bitcoin P2P message types are clearly a different kind of primitives than consensus data structures

I tried splitting this further into `primitives.py` (for the primitive consensus data structures) and `messages.py` (for the p2p message data structures), but the resulting layout didn't make much sense to me so I reverted it. The main motivation for this PR was to split out the mininode functionality from the underlying data structures. I've renamed the new module messages.py as suggested by @ryanofsky.

If you think it's worth splitting further into consensus primitives and p2p message data structures, I can look at doing that in a future PR. It would end up looking something like https://github.com/jnewbery/bitcoin/tree/pr11648.2.

laanwj commented:
@jnewbery Thanks. Calling it messages.py is enough. It was just that having two concepts of "primitives" between the tests and code was confusing to me.

practicalswift commented:
utACK 1135c796a0e1151b03933f2cd36ce95a8e136b7a

MarcoFalke commented on test/functional/test_framework/messages.py:
```diff
@@ -0,0 +1,1319 @@
+MAX_INV_SZ = 50000
+MAX_BLOCK_BASE_SIZE = 1000000
+
+COIN = 100000000 # 1 btc in satoshis
```
nit: This is a conversion helper constant not used by any of the primitives or messages.

laanwj replied:
It's used in:

- `CTxOut.__repr__`
- `CTransaction.is_valid`

It's arguable where this belongs, the ideal place might be some bitcoin-specific constants header, but for now it makes sense to have it here.

MarcoFalke replied:
You are right, I missed those. Looks fine now.

maflcko reviewed (commented):
utACK 1135c796a0e1151b03933f2cd36ce95a8e136b7a

Verified that it is indeed move-only and/or fixups due to rename.
I am neutral on the overall change. Some prefer a huge file where it is easy to jump back and forth, others prefer multiple files that bundle into smaller sets of related classes and functions.

I think there is no single true way of organizing the files, but if the refactoring helps some greater goal, I am happy with the changes.

laanwj commented:
utACK 1135c79

promag commented on test/functional/test_framework/messages.py:
```diff
@@ -0,0 +1,1319 @@
+    return sha256(sha256(s))
+
+def ser_compact_size(l):
+    r = b""
```
Unnecessary?

MarcoFalke replied:
Generally, it is easier to verify move-only-ness when no code changes are made. We've had pull requests in the past that did "all the refactoring" in one go and introduced bugs.

laanwj replied:
IMO both ways are acceptable, as long as the commits that move code are separate from those that make changes.

In any case, moving code around doesn't obligate someone to also fix all the problems it has, in style or other things. Can be done later.

MarcoFalke replied:
Yes, happy to see this fixed up when you touch that function for other reasons. Not sure if this change warrants a pull request on itself.
